lib/ukstore/include/uk/store.h

```diff
@@ -29,3 +29,404 @@
+
+/**
+ * Get a dynamic folder from a library
+ *
+ * @param folder_name the name of the folder to search for
+ * @param library_id the id returned by uk_store_libid
+ * @return the folder searched for
+ */
+#define uk_store_get_folder(library_id, folder_name)	\
+	_uk_store_get_folder(library_id, folder_name)
+
+struct uk_store_folder *
+_uk_store_get_folder(unsigned int library_id, const char *name);
+
+struct uk_store_entry *
+_uk_store_get_dynamic_entry(unsigned int libid, const char *f_name,
+				const char *e_name);
+
+static inline struct uk_store_entry *
+_uk_store_get_entry(unsigned int libid, const char *f_name, const char *e_name)
+{
+	return _uk_store_get_dynamic_entry(libid, f_name, e_name);
+}
+
+/**
+ * Searches for an entry in a folder in a library. Increases the refcount.
+ *
+ * @param libname the name of the library to search in
+ * @param foldername the name of the folder to search in
+ * @param entryname the name of the entry to search for
+ * @return the found entry or NULL
+ */
+#define uk_store_get_entry(libname, foldername, entryname)	\
+	_uk_store_get_entry(uk_store_libid(libname), foldername, entryname)
+
+void
+_uk_store_free_entry(struct uk_store_entry **entry);
+
+/**
+ * Decreases the refcount. When it reaches 0, the memory is freed.
+ *
+ * @param entry the entry to free
+ */
+#define uk_store_free_entry(entry)	\
+	_uk_store_free_entry((entry))
+
+
+/**
+ * Saves a new getter function in the entry
+ *
+ * @param entry the place where to store the function
+ * @param e_type the new type of the function
+ * @param func the function
+ */
+#define uk_store_update_getter(entry, e_type, func)			\
+	do {								\
+		if ((entry)->type != uk_store_get_entry_type(e_type))	\
+			(entry)->set.e_type = NULL;			\
+		(entry)->type = uk_store_get_entry_type(e_type);	\
+		(entry)->get.e_type = func;				\
+	} while (0)
+
+/**
+ * Saves a new setter function in the entry
+ *
+ * @param entry the place where to save the new function
+ * @param e_type the new type of the function
+ * @param func the function
+ */
+#define uk_store_update_setter(entry, e_type, func)			\
+	do {								\
+		if ((entry)->type != uk_store_get_entry_type(e_type))	\
+			(entry)->get.e_type = NULL;			\
+		(entry)->type = uk_store_get_entry_type(e_type);	\
+		(entry)->set.e_type = func;				\
+	} while (0)
+
+int
+_uk_store_get_u8(struct uk_store_entry *e, __u8 *out);
+int
+_uk_store_get_s8(struct uk_store_entry *e, __s8 *out);
+int
+_uk_store_get_u16(struct uk_store_entry *e, __u16 *out);
+int
+_uk_store_get_s16(struct uk_store_entry *e, __s16 *out);
+int
+_uk_store_get_u32(struct uk_store_entry *e, __u32 *out);
+int
+_uk_store_get_s32(struct uk_store_entry *e, __s32 *out);
+int
+_uk_store_get_u64(struct uk_store_entry *e, __u64 *out);
+int
+_uk_store_get_s64(struct uk_store_entry *e, __s64 *out);
+int
+_uk_store_get_uptr(struct uk_store_entry *e, __uptr *out);
+int
+_uk_store_get_charp(struct uk_store_entry *e, char **out);
```

Comment:
I agree, I asked myself this too. But, by changing the signature it will kind of break the design. Maybe it is easier to just add a comment for the getter to ensure that the caller does not give a buffer, but frees the one received afterwards.

Reply:
Right now the user should give a pointer and Ukstore will allocate a buffer to it which needs to be freed afterwards.

Reply:
Do you think it is better to switch to user given buffer+size? (is it worth?)